kliff/trainer/base_trainer.py

```diff
@@ -0,0 +1,636 @@
+                        "kwargs", None
+                    )
+                    if not kwargs:
+                        raise TrainerError(
+                            "Configuration transform module options not provided."
+                        )
+                    configuration_module = configuration_module(
+                        **kwargs, copy_to_config=True
+                    )
+
+                    for config in self.dataset.configs:
+                        _ = configuration_module(config, return_extended_state=True)
+
+                    self.configuration_transform = configuration_module
+        else:
+            logger.warning("No transforms provided. Using raw dataset.")
+
+    def save_config(self):
+        """
+        Hash and save the configuration to the current run directory.
+        """
+        config_hash = self.get_trainer_hash()
+        config_file = f"{self.current['run_dir']}/{config_hash}.yaml"
+        with open(config_file, "w") as f:
+            yaml.dump(self.config_to_dict(), f, default_flow_style=False)
+        logger.info(f"Configuration saved in {config_file}.")
+
+    def setup_model(self):
+        """
+        Set up the model based on the provided information. If the model is not provided,
+        it will be loaded from the model_path. If the model_path is not provided, it will
+        raise an error. If the model_type is KIM, it will be loaded from the KIM model
+        repository. If KIM type model is installed in CWD, it will be loaded from there, and
+        model_path will be set to the KIM CWD. If model is of type TAR, it will be untarred
+        and model_path will be set to the untarred directory. Left for the derived classes
+        to implement.
+        """
+        raise TrainerError("setup_model not implemented.")
+
+    def setup_parameter_transforms(self):
+        """
+        This method set up the transformed parameter space for models. It can be used
+        for any model type in general, but as there exists a significant difference
+        between how models handles their parameters, it is left for the subclass to
+        implement. Although to ensure that `initialize` function remains consistent
+        this method will not raise NotImplemented error, rather it will quietly pass.
+        So be aware.
+        """
+        pass
+
+    def setup_optimizer(self):
+        """
+        Set up the optimizer based on the provided information. If the optimizer is not
+        provided, it will be loaded from the optimizer_name. If the optimizer_name is not
+        provided, it will raise an error. If the optimizer_provider is scipy, it will be
+        loaded from the scipy.optimize. If the optimizer_provider is torch, it will be
+        loaded from the torch.optim. Left for the derived classes to implement.
+        """
+        raise TrainerError("setup_optimizer not implemented.")
+
+    def setup_test_train_datasets(self):
```

Comment:
1. Name should be updated, like `setup_dataset_split`?
2. We only do train and val below; probably add test?

Reply:
Renamed. For test, see comment above.